sdk/tests/deqp/modules/shared/glsBufferTestUtil.js

```diff
@@ -0,0 +1,1068 @@
+    };
+
+    /**
+     * @param {number} offset
+     * @param {number} numBytes
+     * @param {Uint8Array} bytes
+     */
+    glsBufferTestUtil.ReferenceBuffer.prototype.setSubData = function(offset, numBytes, bytes) {
+        assertMsgOptions(deMath.deInBounds32(offset, 0, this.m_data.byteLength) && deMath.deInRange32(offset + numBytes, offset, this.m_data.byteLength),
+            'Parameters not in buffer bounds or range', false, true);
+        var dst = new Uint8Array(this.m_data, offset);
+        dst.set(bytes.subarray(offset, offset + numBytes));
+    };
+
+    // Buffer writer system.
+
+    /**
+     * @enum {number}
+     */
+    glsBufferTestUtil.WriteType = {
+        BUFFER_SUB_DATA: 0,
+        BUFFER_WRITE_MAP: 1,
+        TRANSFORM_FEEDBACK: 2,
+        PIXEL_PACK: 3
+    };
+
+    /**
+     * @param {glsBufferTestUtil.WriteType} write
+     * @return {string}
+     */
+    glsBufferTestUtil.getWriteTypeDescription = function(write) {
+        /** @type {Array<string>} */ var s_desc = [
+            'gl.bufferSubData()',
+            'gl.mapBufferRange()',
+            'transform feedback',
+            'gl.readPixels() into PBO binding'
+        ];
+        return /** @type {string} */ (deUtil.getArrayElement(s_desc, write));
+    };
+
+    // BufferWriterBase
+
+    /**
+     * @constructor
+     */
+    glsBufferTestUtil.BufferWriterBase = function() {};
+
+    /**
+     * //Meant to be overriden
+     * @return {number}
+     */
+    glsBufferTestUtil.BufferWriterBase.prototype.getMinSize = function() { throw new Error('Must be overriden'); };
+
+    /**
+     * //Meant to be overriden
+     * @return {number}
+     */
+    glsBufferTestUtil.BufferWriterBase.prototype.getAlignment = function() { throw new Error('Must be overriden'); };
+
+    /**
+     * //Meant to be overriden
+     * @param {WebGLBuffer} buffer
+     * @param {number} offset
+     * @param {number} numBytes
+     * @param {Uint8Array} bytes
+     */
+    glsBufferTestUtil.BufferWriterBase.prototype.writeNoTarget = function(buffer, offset, numBytes, bytes) { throw new Error('Must be overriden'); };
+
+    /**
+     * @param {WebGLBuffer} buffer
+     * @param {number} offset
+     * @param {number} numBytes
+     * @param {Uint8Array} bytes
+     * @param {number} targetHint
+     */
+    glsBufferTestUtil.BufferWriterBase.prototype.write = function(buffer, offset, numBytes, bytes, targetHint) {
+        this.writeNoTarget(buffer, offset, numBytes, bytes);
+    };
+
+    // BufferWriter
+
+    /**
+     * @constructor
+     * @param {glsBufferTestUtil.WriteType} writeType
+     */
+    glsBufferTestUtil.BufferWriter = function(writeType) {
+        /** @type {glsBufferTestUtil.BufferWriterBase} */ this.m_writer = null;
+        switch (writeType) {
+            case glsBufferTestUtil.WriteType.BUFFER_SUB_DATA: this.m_writer = new glsBufferTestUtil.BufferSubDataWriter(); break;
+            default:
+                testFailed('Unsupported writer');
```

Comment:
Sounds good -- I doubt that WRITE_MAP will be supported in WebGL 2.0.

Will the other ways of writing buffers (transform feedback, pixel pack) be used in subsequent tests?